base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,811 @@
+
+    GetCommonString(DB_SCOPE, tmp);
+    if (tmp.CompareNoCase(L"User") == 0)
+        Info.PerUser = TRUE;
+    else if (tmp.CompareNoCase(L"Machine") == 0)
+        Info.PerUser = FALSE;
+    else
+        Info.PerUser = PerUserModeDefault;
+
+    hr = GetProgramFilesPath(installdir, Info.PerUser, Info.GetGuiOwner());
+    if ((Info.Error = ErrorFromHResult(hr)) != 0)
+        return ErrorBox(Info.Error);
+
+    GetGenerateString(DB_GENINST_DIR, tmp);
+    if (tmp.Find('%') == 0 && ExpandEnvStrings(tmp))
+        installdir = tmp;
+    else if (tmp.Compare(None))
+        installdir = BuildPath(installdir, tmp.IsEmpty() ? AppName : tmp.GetString());
+    Info.InstallDir = installdir.GetString();
+    Info.InstallDirLen = installdir.GetLength() + 1;
+    hr = SHPathPrepareForWriteW(Info.GetGuiOwner(), NULL, installdir, SHPPFW_DIRCREATE);
+    if ((Info.Error = ErrorFromHResult(hr)) != 0)
+        return ErrorBox(Info.Error);
+
+    tempdir.Format(L"%s\\~RAM%u.tmp", installdir.GetString(), GetCurrentProcessId());
+    Info.Error = CreateDirectoryTree(tempdir.GetString());
+    if (Info.Error)
+        return ErrorBox(Info.Error);
+
+    if (!*GetGenerateString(DB_GENINST_LNK, shortcut))
+        shortcut.Format(L"%s.lnk", AppName);
+    Info.ShortcutFile = shortcut.Compare(None) ? shortcut.GetString() : NULL;
+
+    LPCWSTR arpkeyname = AppName;
+    tmp = BuildPath(REGPATH_UNINSTALL, arpkeyname);
+    HKEY hRoot = Info.PerUser ? HKEY_CURRENT_USER : HKEY_LOCAL_MACHINE;
+    REGSAM regsam = KEY_READ | KEY_WRITE | (IsSystem64Bit() ? KEY_WOW64_64KEY : KEY_WOW64_32KEY);
+    Info.Error = arpkey.Create(hRoot, tmp, NULL, REG_OPTION_NON_VOLATILE, regsam);
+    if (!Info.Error)
+        Info.Error = Info.Entries.Create(arpkey, GENERATE_ARPSUBKEY);
+    if (Info.Error)
+        ErrorBox(Info.Error);
+
+    if (!Info.Error)
+    {
+        BOOL isCab = SplitFileAndDirectory(Archive).Right(4).CompareNoCase(L".cab") == 0;
+        Info.Error = isCab ? ExtractCab(Archive, tempdir, ExtractCallback, &Info)
+                           : ExtractZip(Archive, tempdir, ExtractCallback, &Info);
+    }
+
+    if (!Info.Error)
+    {
+        SendMessage(Info.hDlg, IM_STARTPROGRESS, 0, 0);
+
+        for (int pos = 1; pos > 0 && !Info.Error;)
+        {
+            pos = files.Find(L'|');
+            CStringW item = pos <= 0 ? files : files.Left(pos);
+            files = files.Mid(pos + 1);
+            InstallFiles(tempdir, item, installdir);
+        }
+
+        if (!Info.MainApp.IsEmpty())
+        {
+            AddUninstallOperationsFromDB(DB_GENINST_DELREG, UNOP_REGKEY);
+            AddUninstallOperationsFromDB(DB_GENINST_DELREGEMPTY, UNOP_EMPTYREGKEY);
+            AddUninstallOperationsFromDB(DB_GENINST_DELFILE, UNOP_FILE, CStringW(L"\\"));
+            AddUninstallOperationsFromDB(DB_GENINST_DELDIR, UNOP_DIR, CStringW(L"\\"));
+            AddUninstallOperationsFromDB(DB_GENINST_DELDIREMPTY, UNOP_EMPTYDIR, CStringW(L"\\"));
```

Comment:
Isn't it possible to pass a unicode string directly to the AddUninstall... function without having to create that intermediate CStringW?